common/djangoapps/student/views.py

```diff
@@ -855,6 +863,10 @@ def dashboard(request):
     valid_verification_statuses = ['approved', 'must_reverify', 'pending', 'expired']
     display_sidebar_on_dashboard = len(order_history_list) or verification_status in valid_verification_statuses
 
+    # Filter out any course enrollment course cards that are associated with fulfilled entitlements
+    for entitlement in [e for e in course_entitlements if e.enrollment_course_run is not None]:
```

Comment:
Does `e.enrollment_course_run` run an additional query for every entitlement to get the enrollment associated with it?

Reply:
Not in this case, it is just an attribute in the model

Reply:
Doesn't it need to lookup the course_run from the database to populate that attribute?

Reply:
It does, the fix would be building a dictionary that maps the two values, here we have a worse case O(2n) solution, which shouldn't cause performance issues and leads to a far cleaner piece of code. Going to leave as is unless you feel strongly otherwise.

Reply:
This looks like it might be an n+1 query, in which case it might be good to do something like what is suggested here: https://stackoverflow.com/questions/6388668/how-to-avoid-n1-select-in-django